### PR TITLE
Add extra salt to username when generating a user

### PR DESCRIPTION
**Pull request by ianopolous**

### CODE REVIEW

**Comment by cesarosum** [comment]
Ouch that's a lot of files!

File: src/peergos/shared/user/ScryptGenerator.java [hunk 1]
```diff
@@ -11,14 +11,22 @@
 
     @JsProperty
     public final int memoryCost, cpuCost, parallelism, outputBytes;
+    @JsProperty
+    public final String extraSalt;
 
-    public ScryptGenerator(int memoryCost, int cpuCost, int parallelism, int outputBytes) {
+    public ScryptGenerator(int memoryCost, int cpuCost, int parallelism, int outputBytes, String extraSalt) {
         if (memoryCost < MIN_MEMORY_COST)
```
Comment (cesarosum) [inline src/peergos/shared/user/ScryptGenerator.java]: Can we either add a comment or more information to the exception message about MIN_MEMORY_COST? Is it 17 bytes? Why that number?

File: src/peergos/shared/user/UserUtil.java [hunk 2]
```diff
@@ -22,7 +22,7 @@
                                                                SecretGenerationAlgorithm algorithm) {
         if (password.equals(username))
             return Futures.errored(new IllegalStateException("Your password cannot be the same as your username!"));
-        CompletableFuture<byte[]> fut = hasher.hashToKeyBytes(username, password, algorithm);
+        CompletableFuture<byte[]> fut = hasher.hashToKeyBytes(username + algorithm.getExtraSalt(), password, algorithm);
```
Comment (cesarosum) [inline src/peergos/shared/user/UserUtil.java]: For my own understanding, why add the salt to the username and not the password?

File: src/peergos/shared/user/RandomSecretType.java [hunk 3]
```diff
@@ -24,4 +24,9 @@ static RandomSecretType fromCbor(CborObject cbor) {
     public Type getType() {
         return Type.Random;
     }
+
+    @Override
+    public String getExtraSalt() {
```
Comment (cesarosum) [inline src/peergos/shared/user/RandomSecretType.java]: So `RandomSecretType` doesn't use an extra salt?

File: src/peergos/shared/crypto/password/PasswordProtected.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
 public class PasswordProtected {
 
     public static SecretGenerationAlgorithm getDefault() {
-        return new ScryptGenerator(ScryptGenerator.MIN_MEMORY_COST, 8, 1, 32);
+        return new ScryptGenerator(ScryptGenerator.MIN_MEMORY_COST, 8, 1, 32, "");
```
Comment (cesarosum) [inline src/peergos/shared/crypto/password/PasswordProtected.java]: To be clear, we consider "no salt" to be the default behaviour?

File: src/peergos/shared/user/UserUtil.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/shared/user/UserUtil.java]: The username is the salt. We are just extending it.

File: src/peergos/shared/user/ScryptGenerator.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/shared/user/ScryptGenerator.java]: 17 is the memory factor in the [scrypt](https://en.wikipedia.org/wiki/Scrypt) hashing function. It means the calculation will use 2^17 bytes of memory. This number was originally chosen to be sufficiently strong against brute force attacks, and to still take ~1s on a mobile.

File: src/peergos/shared/user/RandomSecretType.java (hunk 3 above)
Comment (ianopolous) [reply src/peergos/shared/user/RandomSecretType.java]: Indeed, it is already random, so we are no improving anything by adding salt.

File: src/peergos/shared/crypto/password/PasswordProtected.java (hunk 4 above)
Comment (ianopolous) [reply src/peergos/shared/crypto/password/PasswordProtected.java]: Current users have no extra salt, just the username as salt. New users after this pr, and existing users once they change their password will use a 32 random byte (converted to hex) salt.

File: src/peergos/shared/user/RandomSecretType.java (hunk 3 above)
Comment (cesarosum) [reply src/peergos/shared/user/RandomSecretType.java]: 👍

File: src/peergos/shared/crypto/password/PasswordProtected.java (hunk 4 above)
Comment (cesarosum) [reply src/peergos/shared/crypto/password/PasswordProtected.java]: Got it 👍

File: src/peergos/shared/user/UserUtil.java (hunk 2 above)
Comment (cesarosum) [reply src/peergos/shared/user/UserUtil.java]: Ah I see.

File: src/peergos/shared/user/ScryptGenerator.java (hunk 1 above)
Comment (cesarosum) [reply src/peergos/shared/user/ScryptGenerator.java]: 👍

**Pull request merged**